File: app/src/main/java/com/example/win/easy/dagger/module/ParserModule.java
package com.example.win.easy.dagger.module;

import com.example.win.easy.parser.RegulationFilenameParser;
import com.example.win.easy.parser.filter.CharSequenceFilterStrategy;
import com.example.win.easy.parser.filter.FilterStrategy;
import com.example.win.easy.parser.interfaces.FilenameParser;
import com.example.win.easy.parser.interfaces.MatcherProxy;
import com.example.win.easy.parser.matchers.ChineseCharacterMatcherProxy;
import com.example.win.easy.parser.matchers.WordMatcherProxy;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

@Module
public class ParserModule {

    @Provides
    @Singleton
    FilenameParser<Character> provideCharacterFilenameParser(){
        List<MatcherProxy<Character>> matcherProxies=new ArrayList<>();
        matcherProxies.add(new WordMatcherProxy());
        matcherProxies.add(new ChineseCharacterMatcherProxy());
        return new RegulationFilenameParser(matcherProxies);
    }

    @Provides
    @Singleton
    FilterStrategy<List<Character>> provideCharSequenceFilterStrategy(){
        return new CharSequenceFilterStrategy();
    }

}

File: app/src/main/java/com/example/win/easy/dagger/module/ParserModule.kt
package com.example.win.easy.dagger.module

import com.example.win.easy.parser.RegulationFilenameParser
import com.example.win.easy.parser.filter.CharSequenceFilterStrategy
import com.example.win.easy.parser.filter.FilterStrategy
import com.example.win.easy.parser.interfaces.FilenameParser
import com.example.win.easy.parser.interfaces.MatcherProxy
import com.example.win.easy.parser.matchers.ChineseCharacterMatcherProxy
import com.example.win.easy.parser.matchers.WordMatcherProxy
import dagger.Module
import dagger.Provides
import java.util.*
import javax.inject.Singleton

@Module
class ParserModule {

    @Provides
    @Singleton
    fun provideCharacterFilenameParser(): FilenameParser<Char> {
        val matcherProxies: MutableList<MatcherProxy<Char>> = ArrayList()
        matcherProxies.add(WordMatcherProxy())
        matcherProxies.add(ChineseCharacterMatcherProxy())
        return RegulationFilenameParser(matcherProxies)
    }

    @Provides
    @Singleton
    fun provideCharSequenceFilterStrategy(): FilterStrategy<List<Char>> = CharSequenceFilterStrategy()
}